### DISCUSSION TITLE
Textual vs prompt_toolkit (although it's 🍐s to 🍎 s comparison)

### DISCUSSION THREAD

**Discussion by Henddher** [Textualize/textual#3254] [Q&A] [answered]
### Hello Textualize folks!
First of all, **Great Library** and **Superb Documentation**. Truly great work 👏 👏 👏 

### Brief background
In the past, I used [prompt_toolkit] mostly for its styled rendering. I didn't try the `App` features _explicitly_. Just basic input and output. For a new project, I am trying Textual and so far, so good.

One thing I found and later heavily relied on was an SSH [prompt_toolkit contrib] backed by asynchssh. With very little code, one can create a multi-user application without any "web" related stuff. In my case, a text-based multi-player game. The game is an ssh server and players just connect to it using the terminal of their choice.

The aforementioned [prompt_toolkit contrib] relies on its `Input` and `Output` streams wired to the ssh server session I/O channel (see [code]).

### Question
I admit I have not browsed Rich's code nor Textual so I do not know how `Input` and `Output` are wired.
Is there some documentation/examples of how that's done?

### Goal
Ideally, building something along the lines of that contrib would allow to create multi-user apps with a significantly richer TUI and with very little "web stuff" to worry about. I saw a few posts from other folks inquiring about telnet/ssh too.

[prompt_toolkit]: https://github.com/prompt-toolkit/python-prompt-toolkit
[prompt_toolkit contrib]: https://github.com/prompt-toolkit/python-prompt-toolkit/tree/master/src/prompt_toolkit/contrib
[code]: https://github.com/prompt-toolkit/python-prompt-toolkit/blob/490cf900188df357611cdc9358256be0e5ce2e16/src/prompt_toolkit/contrib/ssh/server.py#L37-L38

**Comment by davep** [comment, marked as answer]
It's not 100% clear to me what the question is here, but it seems like you're asking how Textual handles input and output and you're asking what code you can look at relating to that? If so, I suspect [this is where you want to start](https://github.com/Textualize/textual/tree/main/src/textual/drivers).

**Reply by Henddher** [reply]
Thank you @davep. 
Yes, that’s part of what I’m looking for.
In prompt_toolkit, input and output can be injected. Even for testing, that’s the typical pattern.

From what I gather, something similar might be possible with Textual. The output as a file-like object but I am not sure about the input yet. Unless, perhaps, a full driver is the key? (just like there’s one for windows, Linux and even web).

More insights and suggestions would be welcome.
